src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -67,68 +81,266 @@ where
-                // We are done, send the final count.
-                if let Some(response_protocol) = response_protocol_handle.borrow_mut().deref_mut() {
-                    response_protocol.send_final_count(row_count);
+        let mut uploader = CopyToS3Uploader::new(sdk_config, sink_connection, "part".into());
+
+        let mut row_count = 0;
+        while let Some(event) = input_handle.next().await {
+            match event {
+                AsyncEvent::Data(_ts, data) => {
+                    for ((row, ()), ts, diff) in data {
+                        if !up_to.less_equal(&ts) {
+                            assert!(diff > 0, "negative accumulation in copy to S3 input");
+                            row_count += u64::try_from(diff).unwrap();
+                            for _ in 0..diff {
+                                match uploader.append_row(&row).await {
+                                    Ok(()) => {}
+                                    Err(e) => {
+                                        send_response(CopyToResponse::Error(e.to_string()));
+                                        return;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                AsyncEvent::Progress(frontier) => {
+                    if PartialOrder::less_equal(&up_to, &frontier) {
+                        match uploader.flush().await {
+                            Ok(()) => {
+                                // We are done, send the final count.
+                                send_response(CopyToResponse::RowCount(row_count));
+                                return;
+                            }
+                            Err(e) => {
+                                send_response(CopyToResponse::Error(e.to_string()));
+                                return;
+                            }
+                        }
+                    }
                 }
             }
         }
     });
 }
 
+// Required state to upload batches to S3
+struct CopyToS3Uploader {
+    // The output description.
+    desc: RelationDesc,
+    // Params to format the data.
+    format: CopyFormatParams<'static>,
+    // The index of the current file.
+    file_index: usize,
+    // The prefix for the file names.
+    file_name_prefix: String,
+    // The s3 bucket.
+    bucket: String,
+    //The path prefix where the files should be uploaded to.
+    path_prefix: String,
+    // The desired file size. A new file upload will be started
+    // when the size exceeds this amount.
+    max_file_size: u64,
+    // The aws sdk config.
+    sdk_config: SdkConfig,
+    // Multi-part uploader for the current file.
+    // Keeping the uploader in an `Option` to later take owned value.
+    current_file_uploader: Option<S3MultiPartUploader>,
+    // Temporary buffer to store the encoded bytes.
+    buf: Vec<u8>,
+}
```

Comment:
These should be doc comments, not regular comments.

Reply:
updated